File: app/web/src/lib/lenders.ts
// The marketplace's lender layer — demo lenders as PUBLISHED PRODUCT POLICIES.
//
// Each lender is a policy config (rate tiers, DBR cap, score floor, amount/tenor
// range), NOT a proprietary model: real lenders never hand over their underwriting
// formula, but they do publish product criteria — that is what an aggregator runs on.
// The math here mirrors affordability.py exactly (annuity → installment → DBR vs the
// SAMA cap) so an offer on screen is the same arithmetic the API serves; it runs
// client-side for instant search UX, like the checkIntegrity mirror in adapters.ts.
//
// All lenders are fictional and clearly labelled illustrative in the UI. The final
// credit decision always belongs to the licensed lender (shown on every offer).

import type { ScoreResult } from './api'

export type ProductType = 'auto' | 'personal' | 'goods'

export interface LenderPolicy {
  id: string
  nameEn: string
  nameAr: string
  kind: 'bank' | 'digital' | 'finance' | 'micro'
  kindEn: string
  kindAr: string
  color: string // brand accent for the card monogram
  products: ProductType[]
  minScore: number
  maxRisk: 'low' | 'medium' | 'high' // most permissive Tabaqa risk band accepted
  dbrCap: number // lender's internal installment cap — may be tighter than SAMA's
  minAmount: number
  maxAmount: number
  minTenor: number
  maxTenor: number
  baseRate: number // APR for low-risk applicants
  mediumSpread: number // added for medium risk (risk-based pricing, visible on screen)
  adminFeePct: number // ≤ 0.01 — SAMA caps the admin fee at min(1%, SAR 5,000)
}

// SAMA Responsible Lending Principles, Circular 46538/99 Ch. IV — the binding
// salary-deduction cap. Every lender policy is clamped to it.
export const SAMA_CAP_EMPLOYEE = 0.3333
export const SAMA_ADMIN_FEE_CEILING_SAR = 5_000
export const SAMA_ADMIN_FEE_CEILING_PCT = 0.01

export const LENDERS: LenderPolicy[] = [
  {
    id: 'waha', nameEn: 'Al Waha Bank', nameAr: 'مصرف الواحة',
    kind: 'bank', kindEn: 'Bank', kindAr: 'بنك', color: '#1d4ed8',
    products: ['auto', 'personal'],
    minScore: 72, maxRisk: 'low', dbrCap: 0.30,
    minAmount: 20_000, maxAmount: 500_000, minTenor: 12, maxTenor: 60,
    baseRate: 0.069, mediumSpread: 0.015, adminFeePct: 0.01,
  },
  {
    id: 'sidra', nameEn: 'Sidra Bank', nameAr: 'بنك السدرة',
    kind: 'bank', kindEn: 'Bank', kindAr: 'بنك', color: '#0e7490',
    products: ['auto', 'personal', 'goods'],
    minScore: 62, maxRisk: 'medium', dbrCap: 0.3333,
    minAmount: 10_000, maxAmount: 300_000, minTenor: 12, maxTenor: 60,
    baseRate: 0.084, mediumSpread: 0.015, adminFeePct: 0.01,
  },
  {
    id: 'mada', nameEn: 'Mada Digital Bank', nameAr: 'بنك المدى الرقمي',
    kind: 'digital', kindEn: 'Digital bank', kindAr: 'بنك رقمي', color: '#7c3aed',
    products: ['auto', 'personal', 'goods'],
    minScore: 55, maxRisk: 'medium', dbrCap: 0.3333,
    minAmount: 5_000, maxAmount: 150_000, minTenor: 6, maxTenor: 48,
    baseRate: 0.099, mediumSpread: 0.02, adminFeePct: 0.005,
  },
  {
    id: 'nakhla', nameEn: 'Nakhla Finance', nameAr: 'النخلة للتمويل',
    kind: 'finance', kindEn: 'Finance co.', kindAr: 'شركة تمويل', color: '#b45309',
    products: ['auto', 'personal', 'goods'],
    minScore: 45, maxRisk: 'medium', dbrCap: 0.3333,
    minAmount: 5_000, maxAmount: 100_000, minTenor: 6, maxTenor: 36,
    baseRate: 0.125, mediumSpread: 0.03, adminFeePct: 0.01,
  },
  {
    id: 'yusr', nameEn: 'Yusr Microfinance', nameAr: 'يُسر للتمويل الأصغر',
    kind: 'micro', kindEn: 'Microfinance', kindAr: 'تمويل أصغر', color: '#0f9d63',
    products: ['personal', 'goods'],
    minScore: 35, maxRisk: 'high', dbrCap: 0.3333,
    minAmount: 2_000, maxAmount: 30_000, minTenor: 6, maxTenor: 24,
    baseRate: 0.15, mediumSpread: 0.04, adminFeePct: 0.01,
  },
]

export interface OfferSearch {
  product: ProductType
  amount: number | null // null → "max I can get" mode
  tenor: number // requested months; clamped per lender
}

export interface Offer {
  lender: LenderPolicy
  amount: number
  reducedFrom: number | null // counter-offer: requested amount didn't fit, this did
  tenor: number
  annualRate: number
  installment: number
  adminFee: number
  totalCost: number // installment × tenor + admin fee
  dbrBefore: number
  dbrAfter: number
  dbrCap: number
  maxFinancing: number // at this lender's policy
  annuityFactor: number
  best: boolean
}

export type LockReason =
  | { kind: 'score'; minScore: number; gap: number; unlockedByRecourse: boolean }
  | { kind: 'risk' }
  | { kind: 'dbr'; maxFinancing: number }
  | { kind: 'amount_range'; min: number; max: number }
  | { kind: 'income' }

export interface LockedOffer {
  lender: LenderPolicy
  reason: LockReason
}

export interface OffersResult {
  offers: Offer[]
  locked: LockedOffer[]
  fullOfferCount: number // offers at the full requested amount (the headline number)
  bestMaxFinancing: number // the largest amount any lender extends
}

// Present-value annuity factor — identical to affordability.annuity_factor.
export function annuityFactor(annualRate: number, months: number): number {
  const n = Math.max(1, Math.round(months))
  const i = annualRate / 12
  if (i === 0) return n
  return (Math.pow(1 + i, n) - 1) / (i * Math.pow(1 + i, n))
}

const RISK_ORDER = { low: 0, medium: 1, high: 2 } as const

export interface OfferInputs {
  income: number // verified monthly income the offers run on
  obligations: number // existing monthly obligations (absolute SAR)
  score: number
  riskFlag: string
  recourseProjected: number | null // recourse projected_score, for unlock chips
}

/** Pull the offer inputs out of a scored profile. `bankOnly` swaps the income lens. */
export function offerInputs(r: ScoreResult, bankOnly = false): OfferInputs {
  const trueIncome = r.income.true_monthly_income
  return {
    income: bankOnly ? r.income.bank_only_income : trueIncome,
    // Obligations are real regardless of which income the lender can see.
    obligations: r.features ? Math.round(r.features.recurring_obligation_load * trueIncome) : 0,
    score: r.tabaqa_score,
    riskFlag: r.risk_flag,
    recourseProjected: r.recourse && !r.recourse.already_prime ? r.recourse.projected_score : null,
  }
}

export function computeOffers(inp: OfferInputs, search: OfferSearch): OffersResult {
  const offers: Offer[] = []
  const locked: LockedOffer[] = []

  for (const lender of LENDERS) {
    if (!lender.products.includes(search.product)) continue

    if (inp.income <= 0) {
      locked.push({ lender, reason: { kind: 'income' } })
      continue
    }
    if (inp.score < lender.minScore) {
      locked.push({
        lender,
        reason: {
          kind: 'score',
          minScore: lender.minScore,
          gap: lender.minScore - inp.score,
          unlockedByRecourse: inp.recourseProjected != null && inp.recourseProjected >= lender.minScore,
        },
      })
      continue
    }
    if (RISK_ORDER[inp.riskFlag as keyof typeof RISK_ORDER] > RISK_ORDER[lender.maxRisk]) {
      locked.push({ lender, reason: { kind: 'risk' } })
      continue
    }

    const cap = Math.min(lender.dbrCap, SAMA_CAP_EMPLOYEE)
    const rate = lender.baseRate + (inp.riskFlag === 'medium' ? lender.mediumSpread : 0)
    const tenor = Math.min(Math.max(search.tenor, lender.minTenor), lender.maxTenor)
    const af = annuityFactor(rate, tenor)

    const maxInstallment = cap * inp.income - inp.obligations
    const maxFinancing = Math.min(Math.max(0, maxInstallment) * af, lender.maxAmount)

    const requested = search.amount
    const amount = requested == null ? maxFinancing : Math.min(requested, maxFinancing, lender.maxAmount)
    if (amount < lender.minAmount) {
      if (requested != null && requested >= lender.minAmount && requested <= lender.maxAmount) {
        locked.push({ lender, reason: { kind: 'dbr', maxFinancing } })
      } else {
        locked.push({ lender, reason: { kind: 'amount_range', min: lender.minAmount, max: lender.maxAmount } })
      }
      continue
    }

    const installment = amount / af
    const adminFee = Math.min(
      amount * lender.adminFeePct,
      amount * SAMA_ADMIN_FEE_CEILING_PCT,
      SAMA_ADMIN_FEE_CEILING_SAR,
    )
    offers.push({
      lender,
      amount: Math.round(amount),
      reducedFrom: requested != null && amount < requested ? requested : null,
      tenor,
      annualRate: rate,
      installment: Math.round(installment),
      adminFee: Math.round(adminFee),
      totalCost: Math.round(installment * tenor + adminFee),
      dbrBefore: inp.obligations / inp.income,
      dbrAfter: (inp.obligations + installment) / inp.income,
      dbrCap: cap,
      maxFinancing: Math.round(maxFinancing),
      annuityFactor: af,
      best: false,
    })
  }

  // Rank: full-amount offers first by total cost (cheapest wins the green frame);
  // counter-offers after, largest amount first. In max mode, largest amount wins.
  offers.sort((a, b) => {
    const aFull = a.reducedFrom == null ? 0 : 1
    const bFull = b.reducedFrom == null ? 0 : 1
    if (search.amount == null) return b.amount - a.amount || a.totalCost - b.totalCost
    if (aFull !== bFull) return aFull - bFull
    return aFull === 0 ? a.totalCost - b.totalCost : b.amount - a.amount
  })
  if (offers.length > 0) offers[0].best = true

  return {
    offers,
    locked,
    fullOfferCount: offers.filter((o) => o.reducedFrom == null).length,
    bestMaxFinancing: offers.reduce((m, o) => Math.max(m, o.maxFinancing), 0),
  }
}

File: app/lenders.py
"""Service ④ — the lender layer: demo lenders as PUBLISHED PRODUCT POLICIES.

Server-side twin of ``web/src/lib/lenders.ts``. The browser runs the same math
client-side for instant search UX (like the checkIntegrity mirror in adapters.ts);
this module is the authority the API serves, so an offer on screen is provably the
same arithmetic a lender gets from ``POST /v1/offers``.

Each lender is a **policy config** — rate tiers, DBR cap, score floor, amount and
tenor range — never a proprietary model. Real lenders don't hand over their
underwriting formula, but they do publish product criteria, and that is what a
marketplace runs on.

The chain, and every step is visible to the applicant (no magic numbers):

    max_installment = min(lender_cap, SAMA_CAP) * income - existing_obligations
    AF              = annuity_factor(rate/12, tenor)          # affordability.py
    max_financing   = min(max(0, max_installment) * AF, lender.max_amount)
    amount          = requested is None ? max_financing : min(requested, max_financing)
    installment     = amount / AF
    admin_fee       = min(amount * fee_pct, amount * 1%, SAR 5,000)   # SAMA ceiling

All lenders here are fictional and labelled illustrative. Offers are
pre-qualifications on SAMA rules — the final credit decision always belongs to the
licensed lender.
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field

from affordability import annuity_factor

# SAMA Responsible Lending Principles for Individuals, Circular 46538/99, Ch. IV —
# the binding salary-deduction cap. Every lender policy is clamped to it.
SAMA_CAP_EMPLOYEE = 0.3333
SAMA_ADMIN_FEE_CEILING_SAR = 5_000.0
SAMA_ADMIN_FEE_CEILING_PCT = 0.01

PRODUCTS = ("auto", "personal", "goods")
RISK_ORDER = {"low": 0, "medium": 1, "high": 2}

DISCLAIMER = (
    "Illustrative lenders with published product policies. Offers are "
    "pre-qualifications computed on SAMA responsible-lending rules — the final "
    "credit decision always belongs to the licensed lender."
)


def _round(x: float) -> float:
    """JS ``Math.round`` semantics (half-up), so Python and TS agree to the riyal."""
    return float(math.floor(x + 0.5))


@dataclass(frozen=True)
class LenderPolicy:
    id: str
    name_en: str
    name_ar: str
    kind: str                       # bank | digital | finance | micro
    products: tuple[str, ...]
    min_score: int
    max_risk: str                   # most permissive Tabaqa risk band accepted
    dbr_cap: float                  # lender's own cap — may be tighter than SAMA's
    min_amount: float
    max_amount: float
    min_tenor: int
    max_tenor: int
    base_rate: float                # annual rate for low-risk applicants
    medium_spread: float            # added for medium risk (risk-based pricing)
    admin_fee_pct: float

    def to_dict(self) -> dict:
        from dataclasses import asdict
        d = asdict(self)
        d["products"] = list(self.products)
        return d


LENDERS: list[LenderPolicy] = [
    LenderPolicy(
        id="waha", name_en="Al Waha Bank", name_ar="مصرف الواحة", kind="bank",
        products=("auto", "personal"),
        min_score=72, max_risk="low", dbr_cap=0.30,
        min_amount=20_000, max_amount=500_000, min_tenor=12, max_tenor=60,
        base_rate=0.069, medium_spread=0.015, admin_fee_pct=0.01,
    ),
    LenderPolicy(
        id="sidra", name_en="Sidra Bank", name_ar="بنك السدرة", kind="bank",
        products=("auto", "personal", "goods"),
        min_score=62, max_risk="medium", dbr_cap=0.3333,
        min_amount=10_000, max_amount=300_000, min_tenor=12, max_tenor=60,
        base_rate=0.084, medium_spread=0.015, admin_fee_pct=0.01,
    ),
    LenderPolicy(
        id="mada", name_en="Mada Digital Bank", name_ar="بنك المدى الرقمي", kind="digital",
        products=("auto", "personal", "goods"),
        min_score=55, max_risk="medium", dbr_cap=0.3333,
        min_amount=5_000, max_amount=150_000, min_tenor=6, max_tenor=48,
        base_rate=0.099, medium_spread=0.02, admin_fee_pct=0.005,
    ),
    LenderPolicy(
        id="nakhla", name_en="Nakhla Finance", name_ar="النخلة للتمويل", kind="finance",
        products=("auto", "personal", "goods"),
        min_score=45, max_risk="medium", dbr_cap=0.3333,
        min_amount=5_000, max_amount=100_000, min_tenor=6, max_tenor=36,
        base_rate=0.125, medium_spread=0.03, admin_fee_pct=0.01,
    ),
    LenderPolicy(
        id="yusr", name_en="Yusr Microfinance", name_ar="يُسر للتمويل الأصغر", kind="micro",
        products=("personal", "goods"),
        min_score=35, max_risk="high", dbr_cap=0.3333,
        min_amount=2_000, max_amount=30_000, min_tenor=6, max_tenor=24,
        base_rate=0.15, medium_spread=0.04, admin_fee_pct=0.01,
    ),
]


@dataclass
class OfferInputs:
    """What the offer engine needs out of a scored profile."""
    income: float                        # the verified monthly income offers run on
    obligations: float = 0.0             # existing monthly obligations (absolute SAR)
    score: int = 0
    risk_flag: str = "medium"
    recourse_projected: int | None = None  # score after recourse steps → unlock hints


@dataclass
class Offer:
    lender_id: str
    lender_name_en: str
    lender_name_ar: str
    lender_kind: str
    amount: float
    reduced_from: float | None           # counter-offer: what was asked for
    tenor_months: int
    annual_rate: float
    installment: float
    admin_fee: float
    total_cost: float                    # installment × tenor + admin fee
    dbr_before: float
    dbr_after: float
    dbr_cap: float
    max_financing: float                 # ceiling under this lender's policy
    annuity_factor: float
    best: bool = False

    def to_dict(self) -> dict:
        from dataclasses import asdict
        return asdict(self)


@dataclass
class LockedOffer:
    """A lender that cannot serve this applicant — and exactly why (the unlock path)."""
    lender_id: str
    lender_name_en: str
    lender_name_ar: str
    reason: str                          # score | risk | dbr | amount_range | income
    detail: dict = field(default_factory=dict)

    def to_dict(self) -> dict:
        from dataclasses import asdict
        return asdict(self)


@dataclass
class Ceiling:
    """The derivation behind the biggest number on screen — nothing is granted.

    income × SAMA cap − obligations = installment room; × the annuity factor = the
    most any lender will extend at this tenor.
    """
    verified_income: float
    obligations: float
    sama_cap: float
    max_installment: float               # regulator-level room, lender-independent
    max_financing: float                 # best across the lenders open to this applicant
    at_lender: str | None
    tenor_months: int

    def to_dict(self) -> dict:
        from dataclasses import asdict
        return asdict(self)


@dataclass
class OffersResult:
    offers: list[Offer]
    locked: list[LockedOffer]
    full_offer_count: int                # offers at the full requested amount (the headline)
    best_max_financing: float
    ceiling: Ceiling

    def to_dict(self) -> dict:
        return {
            "offers": [o.to_dict() for o in self.offers],
            "locked": [l.to_dict() for l in self.locked],
            "full_offer_count": self.full_offer_count,
            "best_max_financing": self.best_max_financing,
            "ceiling": self.ceiling.to_dict(),
        }


def compute_offers(
    inp: OfferInputs,
    product: str = "personal",
    amount: float | None = None,          # None → "the maximum I qualify for"
    tenor_months: int = 48,
) -> OffersResult:
    """Run every lender's published policy against one verified money picture."""
    if product not in PRODUCTS:
        raise ValueError(f"Unknown product '{product}'. Known: {list(PRODUCTS)}")

    offers: list[Offer] = []
    locked: list[LockedOffer] = []
    best_af = 0.0
    ceiling_amount = 0.0
    ceiling_lender: str | None = None

    for lender in LENDERS:
        if product not in lender.products:
            continue

        if inp.income <= 0:
            locked.append(LockedOffer(lender.id, lender.name_en, lender.name_ar, "income"))
            continue
        if inp.score < lender.min_score:
            locked.append(LockedOffer(
                lender.id, lender.name_en, lender.name_ar, "score",
                {"min_score": lender.min_score,
                 "gap": lender.min_score - inp.score,
                 "unlocked_by_recourse": (inp.recourse_projected is not None
                                          and inp.recourse_projected >= lender.min_score)},
            ))
            continue
        if RISK_ORDER.get(inp.risk_flag, 1) > RISK_ORDER[lender.max_risk]:
            locked.append(LockedOffer(lender.id, lender.name_en, lender.name_ar, "risk",
                                      {"max_risk": lender.max_risk}))
            continue

        cap = min(lender.dbr_cap, SAMA_CAP_EMPLOYEE)
        rate = lender.base_rate + (lender.medium_spread if inp.risk_flag == "medium" else 0.0)
        tenor = min(max(int(tenor_months), lender.min_tenor), lender.max_tenor)
        af = annuity_factor(rate / 12.0, tenor)

        max_installment = cap * inp.income - inp.obligations
        max_financing = min(max(0.0, max_installment) * af, lender.max_amount)

        if max_financing > ceiling_amount:
            ceiling_amount, ceiling_lender, best_af = max_financing, lender.id, af

        got = max_financing if amount is None else min(amount, max_financing, lender.max_amount)
        if got < lender.min_amount:
            # Distinguish "you can't afford it here" from "this lender doesn't do that size".
            if amount is not None and lender.min_amount <= amount <= lender.max_amount:
                locked.append(LockedOffer(lender.id, lender.name_en, lender.name_ar, "dbr",
                                          {"max_financing": _round(max_financing)}))
            else:
                locked.append(LockedOffer(lender.id, lender.name_en, lender.name_ar, "amount_range",
                                          {"min_amount": lender.min_amount,
                                           "max_amount": lender.max_amount}))
            continue

        installment = got / af
        admin_fee = min(got * lender.admin_fee_pct,
                        got * SAMA_ADMIN_FEE_CEILING_PCT,
                        SAMA_ADMIN_FEE_CEILING_SAR)
        offers.append(Offer(
            lender_id=lender.id,
            lender_name_en=lender.name_en,
            lender_name_ar=lender.name_ar,
            lender_kind=lender.kind,
            amount=_round(got),
            reduced_from=(_round(amount) if amount is not None and got < amount else None),
            tenor_months=tenor,
            annual_rate=round(rate, 4),
            installment=_round(installment),
            admin_fee=_round(admin_fee),
            total_cost=_round(installment * tenor + admin_fee),
            dbr_before=round(inp.obligations / inp.income, 4),
            dbr_after=round((inp.obligations + installment) / inp.income, 4),
            dbr_cap=cap,
            max_financing=_round(max_financing),
            annuity_factor=round(af, 4),
        ))

    # Rank: full-amount offers first, cheapest total cost wins the "best" badge;
    # counter-offers after, largest amount first. In max mode, largest amount wins.
    if amount is None:
        offers.sort(key=lambda o: (-o.amount, o.total_cost))
    else:
        offers.sort(key=lambda o: (o.reduced_from is not None,
                                   o.total_cost if o.reduced_from is None else -o.amount))
    if offers:
        offers[0].best = True

    return OffersResult(
        offers=offers,
        locked=locked,
        full_offer_count=sum(1 for o in offers if o.reduced_from is None),
        best_max_financing=_round(max((o.max_financing for o in offers), default=0.0)),
        ceiling=Ceiling(
            verified_income=_round(inp.income),
            obligations=_round(inp.obligations),
            sama_cap=SAMA_CAP_EMPLOYEE,
            max_installment=_round(max(0.0, SAMA_CAP_EMPLOYEE * inp.income - inp.obligations)),
            max_financing=_round(ceiling_amount),
            at_lender=ceiling_lender,
            tenor_months=int(tenor_months),
        ),
    )


if __name__ == "__main__":  # the reveal, spoken in offers — Fahd (con_8842)
    verified = OfferInputs(income=10_000, obligations=800, score=82, risk_flag="low")
    bank_only = OfferInputs(income=4_000, obligations=800, score=82, risk_flag="low")

    v = compute_offers(verified, "auto", 60_000, 48)
    b = compute_offers(bank_only, "auto", 60_000, 48)

    print(f"verified 10,000 → {v.full_offer_count} full offers "
          f"(ceiling {v.ceiling.max_financing:,.0f} SAR)")
    for o in v.offers:
        tail = "" if o.reduced_from is None else f"  ← counter-offer (asked {o.reduced_from:,.0f})"
        print(f"  {o.lender_name_en:<22} {o.amount:>9,.0f} SAR  "
              f"{o.installment:>6,.0f}/mo  {o.annual_rate * 100:.1f}%{tail}")
    print(f"bank-only 4,000  → {b.full_offer_count} full offers "
          f"(ceiling {b.ceiling.max_financing:,.0f} SAR)")

    assert b.full_offer_count == 0, "bank-only income must yield no full offers"
    assert v.full_offer_count > b.full_offer_count, "the wallet reveal must add offers"
    assert v.ceiling.max_installment == 2533, v.ceiling.max_installment  # 0.3333×10k − 800
    print("OK — the wallet reveal holds in offers")
